### PR TITLE
Generate configuration files from a template

### PR DESCRIPTION
**Pull request by masci**
### What does this PR do?

 * Adds a `go generate` step to `inv agent.build` and `inv dogstatsd.build` to render the config file from a template
 * Moves the `dist` folder away from the `collector` package

### Motivation

We need to distribute 2 configuration files, one for the agent and another one for dogstastd. The dogstatsd one contains a subset of the settings available in the agent, so we cannot simply rename `datadog.yaml` to something else and distribute it with the dsd package.

At the same time, we don't want to maintain 2 different but very similar config files, that would be very error prone (see @xvello comment https://github.com/DataDog/datadog-agent/pull/771#pullrequestreview-74763728 for more context).

With this approach we only need to maintain one template and different configuration files will be generated at build time.

### CODE REVIEW

File: pkg/config/config_template.yaml [hunk 1]
```diff
@@ -1,3 +1,4 @@
+{{ if .Common }}
```
Comment (olivielpeau) [inline pkg/config/config_template.yaml]: nit: does it make sense to have a `Common` flag? :) (by definition I would expect the `Common` part to always be there)

**Review by olivielpeau — COMMENTED** [review]
one nit + the `forwarder` and `histogram_` options should be in the common part

LGTM overall, not very familiar with the `go generate` tool so I'll let @hush-hush give it a proper review :)

**Review by hush-hush — APPROVED** [review]

File: pkg/config/config_template.yaml [hunk 2]
```diff
@@ -23,7 +24,8 @@ api_key:
 
 # Force the hostname to whatever you want. (default: auto-detected)
 # hostname: mymachine.mydomain
-
+{{ end }}
+{{- if .Agent }}
```
Comment (hush-hush) [inline pkg/config/config_template.yaml]: dogstatsd use goexpvar. we should move `expvar_port` the common part. Same for `forwarder_timeout`, `forwarder_retry_queue_max_size`. `histogram_aggregates` and `histogram_percentiles`.

File: pkg/config/render_config.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+			Autodiscovery:     true,
+			DockerTagging:     true,
+			KubernetesTagging: true,
+			ECS:               true,
+		}
+	case "dogstatsd":
+		return context{
+			Common:        true,
+			Dogstatsd:     true,
+			DockerTagging: true,
```
Comment (hush-hush) [inline pkg/config/render_config.go]: we should also activate the logging part.

**Review by hush-hush — CHANGES_REQUESTED** [review]

File: pkg/config/config_template.yaml [hunk 4]
```diff
@@ -154,12 +160,14 @@ metadata_collectors:
 #     template_url: 127.0.0.1
 #     username:
 #     password:
-
+{{ end -}}
+{{- if .Logging }}
```
Comment (gmmeyer) [inline pkg/config/config_template.yaml]: Why isn't logging in common?

**Review by tmichelet — COMMENTED** [review]
Not sure Ramen's review is relevant in this case, but LGTM

File: pkg/config/config_template.yaml (hunk 2 above)
Comment (masci) [reply pkg/config/config_template.yaml]: dsd has its own `dogstatsd_stats_port` that got lost, re-added and moving the other settings

File: pkg/config/config_template.yaml (hunk 1 above)
Comment (masci) [reply pkg/config/config_template.yaml]: You might want to build an executable only running the metadata collection (let's say a standalone tagger) that needs logging but no api_key. Happy to change the name to something more meaningful.

File: pkg/config/config_template.yaml (hunk 4 above)
Comment (masci) [reply pkg/config/config_template.yaml]: see https://github.com/DataDog/datadog-agent/pull/799#discussion_r150495530

File: pkg/config/render_config.go (hunk 3 above)
Comment (masci) [reply pkg/config/render_config.go]: done

**Comment by mfpierre** [comment]
@masci there's a bunch of rename in this PR that does not match the new subdir per check structure ⚠️

File: .gitignore [hunk 5]
```diff
@@ -20,3 +20,7 @@ bin/
 
 #file generated at build time by message compiler
 agentmsg.h
+
+# go-generated files
+datadog.yaml
+dogstatsd.yaml
```
Comment (mfpierre) [inline .gitignore]: You could use that "insert final new line" setting in your editor 😉

**Review by mfpierre — DISMISSED** [review]

File: pkg/config/render_config.go (hunk 3 above)
Comment (xvello) [reply pkg/config/render_config.go]: DSD's tagger also uses KubernetesTagging and ECS

File: .gitignore (hunk 5 above)
Comment (masci) [reply .gitignore]: `"files.insertFinalNewline": true,` VS code doesn't cease to amaze me 😄

**Review by hush-hush — APPROVED** [review]

**Pull request merged**